File: distributor-service/src/main/java/org/pops/et4/jvm/project/distributor/kafka/KafkaProducerService.java
package org.pops.et4.jvm.project.distributor.kafka;

import org.pops.et4.jvm.project.schemas.events.*;
import org.pops.et4.jvm.project.schemas.repositories.distributor.*;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.kafka.core.KafkaTemplate;
import org.springframework.stereotype.Service;

import java.time.Instant;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;

@Service(KafkaProducerService.BEAN_NAME)
public class KafkaProducerService {

    public static final String BEAN_NAME = "distributorServiceKafkaProducerService";

    private final KafkaTemplate<String, Object> kafkaTemplate;
    private final DistributorRepository distributorRepository;
    private final DistributedGameRepository distributedGameRepository;
    private final PlayerRepository playerRepository;
    private final OwnedGameRepository ownedGameRepository;
    private final ReviewRepository reviewRepository;

    @Autowired
    public KafkaProducerService(
            @Qualifier(KafkaConfig.KAFKA_TEMPLATE_BEAN_NAME) KafkaTemplate<String, Object> kafkaTemplate,
            @Qualifier(DistributorRepository.BEAN_NAME) DistributorRepository distributorRepository,
            @Qualifier(DistributedGameRepository.BEAN_NAME) DistributedGameRepository distributedGameRepository,
            @Qualifier(PlayerRepository.BEAN_NAME) PlayerRepository playerRepository,
            @Qualifier(OwnedGameRepository.BEAN_NAME) OwnedGameRepository ownedGameRepository,
            @Qualifier(ReviewRepository.BEAN_NAME) ReviewRepository reviewRepository
    ) {
        this.kafkaTemplate = kafkaTemplate;
        this.distributorRepository = distributorRepository;
        this.distributedGameRepository = distributedGameRepository;
        this.playerRepository = playerRepository;
        this.ownedGameRepository = ownedGameRepository;
        this.reviewRepository = reviewRepository;
    }

    public void sendExampleEvent(String payload) {
        String topic = ExampleEvent.TOPIC;
        String key = UUID.randomUUID().toString();
        ExampleEvent event = ExampleEvent.newBuilder()
                .setPayload(payload)
                .build();

        CompletableFuture<?> future = this.kafkaTemplate.send(topic, key, event);

        future.whenComplete((result, ex) -> System.out.println("[Producer] " + topic + "(" + key + "): " + (ex==null ? result : ex.getMessage())));
    }

    public void sendGameDistributed(Long distributorId, Long gameId, String gameName) {
        String topic = GameDistributed.TOPIC;
        String key = UUID.randomUUID().toString();
        GameDistributed event = GameDistributed.newBuilder()
                .setDistributorId(distributorId)
                .setGameId(gameId)
                .setGameName(gameName)
                .build();

        CompletableFuture<?> future = this.kafkaTemplate.send(topic, key, event);

        future.whenComplete((result, ex) -> System.out.println("[Producer] " + topic + "(" + key + "): " + (ex==null ? result : ex.getMessage())));
    }

    public void sendPatchDistributed(Long distributorId, Long gameId, String newVersion, String gameName) {
        String topic = PatchDistributed.TOPIC;
        String key = UUID.randomUUID().toString();
        PatchDistributed event = PatchDistributed.newBuilder()
                .setDistributorId(distributorId)
                .setGameId(gameId)
                .setNewVersion(newVersion)
                .setGameName(gameName)
                .build();

        CompletableFuture<?> future = this.kafkaTemplate.send(topic, key, event);

        future.whenComplete((result, ex) -> System.out.println("[Producer] " + topic + "(" + key + "): " + (ex==null ? result : ex.getMessage())));
    }

    public void sendSaleStarted(Long distributorId, Long gameId, Float salePercentage, String gameName) {
        String topic = SaleStarted.TOPIC;
        String key = UUID.randomUUID().toString();
        SaleStarted event = SaleStarted.newBuilder()
                .setDistributorId(distributorId)
                .setGameId(gameId)
                .setSalePercentage(salePercentage)
                .setGameName(gameName)
                .build();

        CompletableFuture<?> future = this.kafkaTemplate.send(topic, key, event);

        future.whenComplete((result, ex) -> System.out.println("[Producer] " + topic + "(" + key + "): " + (ex==null ? result : ex.getMessage())));
    }

    public void sendSendGameFile(Long targetId, Long gameId, String version, String gameName, String platform, String playerName) {
        String topic = SendGameFile.TOPIC;
        String key = UUID.randomUUID().toString();
        SendGameFile event = SendGameFile.newBuilder()
                .setTargetId(targetId)
                .setGameId(gameId)
                .setVersion(version)
                .setGameName(gameName)
                .setPlatform(platform)
                .setPlayerName(playerName)
                .build();

        CompletableFuture<?> future = this.kafkaTemplate.send(topic, key, event);

        future.whenComplete((result, ex) -> System.out.println("[Producer] " + topic + "(" + key + "): " + (ex==null ? result : ex.getMessage())));
    }

    public void sendGameReviewed(Long reviewId, Long distributorId, Integer rating, String comment, Instant publicationDate, java.util.List<Long> positiveReactionPlayerIds, java.util.List<Long> negativeReactionPlayerIds) {
        String topic = GameReviewed.TOPIC;
        String key = UUID.randomUUID().toString();
        GameReviewed event = GameReviewed.newBuilder()
                .setReviewId(reviewId)
                .setDistributorId(distributorId)
                .setRating(rating)
                .setComment(comment)
                .setPublicationDate(publicationDate)
                .setPositiveReactionPlayerIds(positiveReactionPlayerIds)
                .setNegativeReactionPlayerIds(negativeReactionPlayerIds)
                .build();

        CompletableFuture<?> future = this.kafkaTemplate.send(topic, key, event);

        future.whenComplete((result, ex) -> System.out.println("[Producer] " + topic + "(" + key + "): " + (ex==null ? result : ex.getMessage())));
    }

    public void sendReviewRefused(Long reviewId, String playerName, String gameName) {
        String topic = ReviewRefused.TOPIC;
        String key = UUID.randomUUID().toString();
        ReviewRefused event = ReviewRefused.newBuilder()
                .setReviewId(reviewId)
                .setPlayerName(playerName)
                .setGameName(gameName)
                .build();

        CompletableFuture<?> future = this.kafkaTemplate.send(topic, key, event);

        future.whenComplete((result, ex) -> System.out.println("[Producer] " + topic + "(" + key + "): " + (ex==null ? result : ex.getMessage())));
    }

    public void sendCrashReported(Long distributorId, Long gameId, Platform platform, String installedVersion, Integer errorCode, String message) {
        String topic = CrashReported.TOPIC;
        String key = UUID.randomUUID().toString();
        CrashReported event = CrashReported.newBuilder()
                .setDistributorId(distributorId)
                .setGameId(gameId)
                .setPlatform(platform)
                .setInstalledVersion(installedVersion)
                .setErrorCode(errorCode)
                .setMessage(message)
                .build();

        CompletableFuture<?> future = this.kafkaTemplate.send(topic, key, event);

        future.whenComplete((result, ex) -> System.out.println("[Producer] " + topic + "(" + key + "): " + (ex==null ? result : ex.getMessage())));
    }

    public void sendSendPlayerPage(String page) {
        String topic = SendPlayerPage.TOPIC;
        String key = UUID.randomUUID().toString();
        SendPlayerPage event = SendPlayerPage.newBuilder()
                .setPage(page)
                .build();

        CompletableFuture<?> future = this.kafkaTemplate.send(topic, key, event);

        future.whenComplete((result, ex) -> System.out.println("[Producer] " + topic + "(" + key + "): " + (ex==null ? result : ex.getMessage())));
    }

}

File: player-service/src/main/kotlin/org/pops/et4/jvm/project/player/kafka/KafkaProducerService.kt
package org.pops.et4.jvm.project.player.kafka

import org.pops.et4.jvm.project.schemas.events.*
import org.pops.et4.jvm.project.schemas.repositories.player.InstalledGameRepository
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.kafka.core.KafkaTemplate
import org.springframework.stereotype.Service
import java.time.Instant
import java.util.UUID

@Service(KafkaProducerService.BEAN_NAME)
class KafkaProducerService(
    @Qualifier(KafkaConfig.KAFKA_TEMPLATE_BEAN_NAME)
    private val kafkaTemplate: KafkaTemplate<String, Any>,
    @Qualifier(InstalledGameRepository.BEAN_NAME)
    private val installedGameRepository: InstalledGameRepository
) {
	companion object {
        const val BEAN_NAME = "playerServiceKafkaProducerService"
        
        // Valid platform values
        val VALID_PLATFORMS = setOf(
            "WINDOWS", "MACOS", "LINUX", 
            "PS5", "PS4", 
            "XBOX_SERIES", "XBOX_ONE", 
            "SWITCH2", "SWITCH"
        )
    }


    fun sendRegisterPlayer(
        distributorId: Long,
        pseudo: String,
        firstName: String,
        lastName: String,
        birthDate: Instant
    ) {
        val topic = RegisterPlayer.TOPIC
        val key = UUID.randomUUID().toString()
        val event = RegisterPlayer.newBuilder()
            .setDistributorId(distributorId)
            .setPseudo(pseudo)
            .setFirstName(firstName)
            .setLastName(lastName)
            .setBirthDate(birthDate)
            .build()

        val future = kafkaTemplate.send(topic, key, event)

        future.whenComplete { result, ex ->
            val message = ex?.message ?: result
            println("[Producer] $topic($key): $message")
        }
    }

    fun sendPurchaseGame(playerId: Long, gameId: Long) {
        val topic = PurchaseGame.TOPIC
        val key = UUID.randomUUID().toString()
        val event = PurchaseGame.newBuilder()
            .setPlayerId(playerId)
            .setGameId(gameId)
            .build()

        val future = kafkaTemplate.send(topic, key, event)

        future.whenComplete { result, ex ->
            val message = ex?.message ?: result
            println("[Producer] $topic($key): $message")
        }
    }

    fun sendReviewGame(playerId: Long, gameId: Long, rating: Int, comment: String?) {
        // Validate rating (0-5)
        if (rating !in 0..5) {
            println("[Error] Invalid rating: $rating. Rating must be between 0 and 5.")
            return
        }
        
        val topic = ReviewGame.TOPIC
        val key = UUID.randomUUID().toString()
        val event = ReviewGame.newBuilder()
            .setPlayerId(playerId)
            .setGameId(gameId)
            .setRating(rating)
            .setComment(comment)
            .build()

        val future = kafkaTemplate.send(topic, key, event)

        future.whenComplete { result, ex ->
            val message = ex?.message ?: result
            println("[Producer] $topic($key): $message")
        }
    }

    fun sendInstallGame(playerId: Long, gameId: Long, platform: String) {
        // Validate platform
        if (platform !in VALID_PLATFORMS) {
            println("[Error] Invalid platform: $platform. Valid platforms: ${VALID_PLATFORMS.joinToString(", ")}")
            return
        }
        
        // Send Kafka event
        val topic = InstallGame.TOPIC
        val key = UUID.randomUUID().toString()
        val event = InstallGame.newBuilder()
            .setPlayerId(playerId)
            .setGameId(gameId)
            .setPlatform(platform)
            .build()

        val future = kafkaTemplate.send(topic, key, event)

        future.whenComplete { result, ex ->
            val message = ex?.message ?: result
            println("[Producer] $topic($key): $message")
        }
    }

    fun sendUpdateGame(playerId: Long, gameId: Long, platform: String, installedVersion: String) {
        // Validate platform
        if (platform !in VALID_PLATFORMS) {
            println("[Error] Invalid platform: $platform. Valid platforms: ${VALID_PLATFORMS.joinToString(", ")}")
            return
        }
        
        // Get installed game version from database
        val installedGames = installedGameRepository.findAll()
            .filter { it.playerId == playerId && it.gameId == gameId && it.platform?.toString() == platform }
        
        if (installedGames.isEmpty()) {
            println("[Error] Game not installed. Cannot update.")
            println("        Use 'install' command to install the game first.")
            return
        }
        
        val currentInstalledVersion = installedGames.first().installedVersion ?: "1.0.0"
        
        // Send Kafka event with the installed version
        val topic = UpdateGame.TOPIC
        val key = UUID.randomUUID().toString()
        val event = UpdateGame.newBuilder()
            .setPlayerId(playerId)
            .setGameId(gameId)
            .setPlatform(platform)
            .setInstalledVersion(currentInstalledVersion)
            .build()

        val future = kafkaTemplate.send(topic, key, event)

        future.whenComplete { result, ex ->
            val message = ex?.message ?: result
            println("[Producer] $topic($key): $message")
        }
    }

    fun sendUninstallGame(playerId: Long, gameId: Long, platform: String, comment: String?) {
        // Validate platform
        if (platform !in VALID_PLATFORMS) {
            println("[Error] Invalid platform: $platform. Valid platforms: ${VALID_PLATFORMS.joinToString(", ")}")
            return
        }
        
        // Update database: Remove from installed games
        try {
            val installedGames = installedGameRepository.findAll()
                .filter { it.playerId == playerId && it.gameId == gameId && it.platform?.toString() == platform }
            
            if (installedGames.isNotEmpty()) {
                installedGameRepository.deleteAll(installedGames)
                println("[Database] Uninstalled game: Player=${playerId}, Game=${gameId}, Platform=${platform}")
            } else {
                println("[Database] Warning: Game not found in installed games")
            }
        } catch (e: Exception) {
            println("[Database] Error uninstalling game: ${e.message}")
        }
        
        // Send Kafka event
        val topic = UninstallGame.TOPIC
        val key = UUID.randomUUID().toString()
        val event = UninstallGame.newBuilder()
            .setPlayerId(playerId)
            .setGameId(gameId)
            .setPlatform(platform)
            .setComment(comment)
            .build()

        val future = kafkaTemplate.send(topic, key, event)

        future.whenComplete { result, ex ->
            val message = ex?.message ?: result
            println("[Producer] $topic($key): $message")
        }
    }

    fun sendAddPlayTime(playerId: Long, gameId: Long, time: Long) {
        val topic = AddPlayTime.TOPIC
        val key = UUID.randomUUID().toString()
        val event = AddPlayTime.newBuilder()
            .setPlayerId(playerId)
            .setGameId(gameId)
            .setTime(time)
            .build()

        val future = kafkaTemplate.send(topic, key, event)

        future.whenComplete { result, ex ->
            val message = ex?.message ?: result
            println("[Producer] $topic($key): $message")
        }
    }

    fun sendReportCrash(
        playerId: Long,
        gameId: Long,
        platform: String,
        installedVersion: String,
        errorCode: Long,
        message: String
    ) {
        // Validate platform
        if (platform !in VALID_PLATFORMS) {
            println("[Error] Invalid platform: $platform. Valid platforms: ${VALID_PLATFORMS.joinToString(", ")}")
            return
        }
        
        val topic = ReportCrash.TOPIC
        val key = UUID.randomUUID().toString()
        val event = ReportCrash.newBuilder()
            .setPlayerId(playerId)
            .setGameId(gameId)
            .setPlatform(platform)
            .setInstalledVersion(installedVersion)
            .setErrorCode(errorCode)
            .setMessage(message)
            .build()

        val future = kafkaTemplate.send(topic, key, event)

        future.whenComplete { result, ex ->
            val msg = ex?.message ?: result
            println("[Producer] $topic($key): $msg")
        }
    }

    fun sendAddWishedGame(playerId: Long, gameId: Long) {
        val topic = AddWishedGame.TOPIC
        val key = UUID.randomUUID().toString()
        val event = AddWishedGame.newBuilder()
            .setPlayerId(playerId)
            .setGameId(gameId)
            .build()

        val future = kafkaTemplate.send(topic, key, event)

        future.whenComplete { result, ex ->
            val message = ex?.message ?: result
            println("[Producer] $topic($key): $message")
        }
    }

    fun sendRemoveWishedGame(playerId: Long, gameId: Long) {
        val topic = RemoveWishedGame.TOPIC
        val key = UUID.randomUUID().toString()
        val event = RemoveWishedGame.newBuilder()
            .setPlayerId(playerId)
            .setGameId(gameId)
            .build()

        val future = kafkaTemplate.send(topic, key, event)

        future.whenComplete { result, ex ->
            val message = ex?.message ?: result
            println("[Producer] $topic($key): $message")
        }
    }

    fun sendReactReview(playerId: Long, reviewId: Long, reactType: Int) {
        // Validate reactType (0-2: 0=NOTHING, 1=POSITIVE, 2=NEGATIVE)
        if (reactType !in 0..2) {
            println("[Error] Invalid reactType: $reactType. ReactType must be 0=NOTHING, 1=POSITIVE, or 2=NEGATIVE.")
            return
        }
        
        val topic = ReactReview.TOPIC
        val key = UUID.randomUUID().toString()
        val event = ReactReview.newBuilder()
            .setPlayerId(playerId)
            .setReviewId(reviewId)
            .setReactType(reactType)
            .build()

        val future = kafkaTemplate.send(topic, key, event)

        future.whenComplete { result, ex ->
            val message = ex?.message ?: result
            println("[Producer] $topic($key): $message")
        }
    }
}